ONSdigital/census-contact-centre-service
# Cr 996

# Motivation and Context

The CR-996 card describes the problem as follows:

The GET /cases/uprn endpoint fabricates a case to return to Serco in the event that AIMS has the address but RM does not.

There is a field, ceOrgName in the response to Serco which we cannot currently populate when the case is created from an AIMS address.

That field will be added to the AIMs API response (TBC), and it should be mapped into the case we send to Serco, the cached case and the NEW_ADDRESS_REPORTED event we send to RM.

# What has changed

Changes have been made, mainly to the CCSvcBeanMapper, so that the getCasesByUprn endpoint will:
1. return (to Serco) the value of the ceOrgName field from RM if it is found to be populated.
2. set the value of the ceOrgName in a new CachedCase, if need be, amd return it to Serco
3. return (to Serco) the value of the ceOrgName field from an existing CachedCase if it is found to be populated.
4. send RM the value of ceOrgName in the NEW_ADDRESS_REPORTED event

Relevant JUnit tests have also been amended to test the organisation name mappings etc.

# How to test?

Need to make sure that the getCasesByUprn endpoint will:
1. return (to Serco) the value of the ceOrgName field from RM if it is found to be populated.
2. set the value of the ceOrgName in a new CachedCase, if need be, amd return it to Serco
3. return (to Serco) the value of the ceOrgName field from an existing CachedCase if it is found to be populated.
4. send RM the value of ceOrgName in the NEW_ADDRESS_REPORTED event

Manually test these as follows:
1. Open the GCP console and go to the Firestore collection named census-rh-<project name>-new-case. Delete any existing case that has uprn 1347459999 or uprn 10013671754.
2. Check that the mock case service contains a case that has uprn 1347459999 and has a value in its organisationName field (in cases.yml).
3. Run rabbitmq locally e.g. go into the census-rh-service repo, in the terminal, and enter this command: docker-compose up -d
Also, go into the rabbit console and purge the queue named event.case.address.update
4. Run the following command to make sure that you have permission to access Firestore (no error occurs otherwise - it just freezes and fails): gcloud auth application-default login
5. Run the mock case service and the ccsvc in eclipse
6. In postman run http://localhost:8171/cases/uprn/1347459999
7. Confirm test 1; that the value of organisationName from "RM" appears in the ceOrgName field that gets returned to postman.
Also, make sure that no new case has been created in Firestore this time.
8. In postman run http://localhost:8171/cases/uprn/10013671754
9. Make sure that the value of organisationName from AI appears in the ceOrgName field that gets returned to postman. NB. in AI we call this:

https://rh-dev-ai-api.ai.census-gcp.onsdigital.uk/addresses/rh/uprn/10013671754?addresstype=paf&verbose=true

And get this:
{"apiVersion":"1.0.0-SNAPSHOT","dataVersion":"72","response":{"address":{"uprn":"10013671754","formattedAddress":"Gorseway Retirement & Nursing Home, 354 Sea Front, Hayling Island, PO11 0BA","addressLine1":"Gorseway Retirement & Nursing Home","addressLine2":"354 Sea Front","addressLine3":"","townName":"Hayling Island","postcode":"PO11 0BA","foundAddressType":"PAF","censusAddressType":"CE","censusEstabType":"Care Home","countryCode":"E",
"organisationName":"Gorseway Retirement & Nursing Home"},"addressType":"PAF","historical":true,"epoch":""},"status":{"code":200,"message":"Ok"},"errors":[]}

And in postman we get this:
[
    {
        "id": "3bb17269-6007-4fba-8b53-f4148cee8842",
        "caseRef": null,
        "caseType": "CE",
        "addressType": "CE",
        "estabType": "CARE_HOME",
        "estabDescription": "Care Home",
        "allowedDeliveryChannels": [
            "POST",
            "SMS"
        ],
        "createdDateTime": "2020-06-26T10:58:39.758Z",
        "lastUpdated": null,
        "addressLine1": "Gorseway Retirement & Nursing Home",
        "addressLine2": "354 Sea Front",
        "addressLine3": "",
        "townName": "Hayling Island",
        "region": "E",
        "postcode": "PO11 0BA",
        "ceOrgName": "Gorseway Retirement & Nursing Home",
        "uprn": 10013671754,
        "estabUprn": null,
        "caseEvents": null,
        "secureEstablishment": false
    }   
]

i.e. the ceOrgName value is populated. 

10. Now in Firestore change the value in the ceOrgName field. To confirm test 3: check that the new value gets returned if you call the same postman endpoint again:
http://localhost:8171/cases/uprn/10013671754

11. Finally test 4; go to rabbit management console and check the queue named event.case.address.update to see if it contains a NEW_ADDRESS_REPORTED event. If so then
check that the collectionCase json contains a value named organisationName, which is populated. NB. Currently there is an organisationName json field but it is null.
